sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala

```diff
@@ -47,4 +48,9 @@ case object RightOuter extends JoinType
 
 case object FullOuter extends JoinType
 
-case object LeftSemi extends JoinType
+abstract class LeftSemiJoin extends JoinType
```

Comment:
Making a common abstract parent class for `LeftSemiJoin` and `LeftAntiJoin` will reduce the code change in `Optimizer` and `Strategies`, but you're right, I didn't give it a proper name to reflect the concept, not sure if you have better idea.